### ISSUE TITLE
[Bug Report] Bloom - bad accuracy of output for Q and A task

### ISSUE THREAD

**Issue by jliangTT** [tenstorrent/tt-metal#7389]
**Describe the bug**

Bloom PCC issue required debug

**To Reproduce**

Status from @arakhmati - 

> I wasn't able to reproduce the bloom linear failure in a standalone test. After looking into it more, I realized that all of the ops match their local golden pretty well. But the global golden gets worse over time. My hunch is that it's because we use bfloat8_b in the optimized bloom and we slowly diverge from torch.  

Next step: Can you or MCW run that test with only bfloat16

**Expected behavior**

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit)

**Additional context**

**Comment by punithsekar** [comment]
The ttnn_optimized_bloom which is available in main has intermediate tensors as dtype=bfloat16 mostly. Now, I kept all the intermediate tensors to dtype=bfloat16 which improved the accuracy slightly.

The observations are noted on the [sheet](https://docs.google.com/spreadsheets/d/1ZmEkZ0Rs8muHBHhe2apmM_1SuDDbLc9gJPZLpey40VU/edit#gid=0) and the code change made for dtype16 is available in this [commit](https://github.com/tenstorrent-metal/tt-metal/compare/main...punith/bloom_qa_with_different_dtype).

**Comment by yugaoTT** [comment]
which cmd should I run to repro?

**Comment by punithsekar** [comment]
To reproduce the pcc issue,  run` tests/ttnn/integration_tests/bloom/test_bloom_for_question_answering.py` test with [sheet](https://docs.google.com/spreadsheets/d/1ZmEkZ0Rs8muHBHhe2apmM_1SuDDbLc9gJPZLpey40VU/edit#gid=0) respective input. This helps to reproduce issue on main.

To reproduce results of bloom with float16 use the following [commit](https://github.com/tenstorrent-metal/tt-metal/compare/main...punith/bloom_qa_with_different_dtype).

**Comment by yugaoTT** [comment]
the LN seems to not be the problem, after switching ever LN to torch, I got 
`start  (False, 'Max ATOL Delta: 14.176633834838867, Max RTOL Delta: 1.2146252393722534, 0.8624867817780981')
end  (False, 'Max ATOL Delta: 11.56474781036377, Max RTOL Delta: 32.350425720214844, 0.8353366016316789')
ttnn answer: `

for the question "Who ruled the duchy of Normandy"

It is the same output as the ttnn LN

**Comment by yugaoTT** [comment]
for better accuracy, you can disable math approx mode for LN and SMX, this can reduce the accumulation error, but with the cost of perf decrease

the disabled version is in this commit `7368d85e9624e1cda2879c2d02618ce046f39b1b`

**Comment by boris-drazic** [comment]
@punithsekar
Bloom demo does not seem to be well suited for the model we are using. Both TT and PyTorch models are producing bad answers (or no answers) with PyTorch being slightly better but not good.
Maybe a simple way to improve Bloom output is to change how we are using Bloom.

Can you rework demo for Bloom Q and A so that it is doing open quesitons, that is, instead of offering answer from context it generates its own answer? So the prompt would be something like
`prompt = "Given the question delimited by double square brackets [[ What is the age of the universe? ]], what is the answer? Answer:"`
For this we would like to see PyTorch model answers and also TT answers to compare them.

Second thing to try is to review Bloom for text genration demo (`models/experimental/functional_bloom/demo/demo.py`) and compare what it produces to PyTorch output.

**Comment by boris-drazic** [comment]
Results for [text generation](https://docs.google.com/spreadsheets/d/1Te8WrCybaw45zcd1RavCj9a_lmBy5Zwl9vdTkzFKzNU/edit#gid=0) and [question answering](https://docs.google.com/spreadsheets/d/1-K_Y8Z0TwJY7WwNXMTw1iE2RUx1zFdUcKcu7eu6QTKI/edit#gid=0) are looking good.

To narrow down best setup for model, can you generate results for ~10 prompts for each of the following and also record the performance like tokens / second:

- `bfloat8_b` accumulation, split linear and gelu instead of fused op, compute_attention_probs in ttnn with `add` and `softmax` (not fused), max_length=20

- `bfloat16` accumulation, split linear and gelu instead of fused op, compute_attention_probs in ttnn with `add` and `softmax` (not fused), max_length=20

- `bfloat16` accumulation, split linear and gelu instead of fused op, compute_attention_probs in PyTorch, max_length=20

- `bfloat8_b` accumulation, split linear and gelu instead of fused op, compute_attention_probs in PyTorch, max_length=20

**Comment by boris-drazic** [comment]
accuracy is fine for text generation and question answering
